filament/src/ShadowMapManager.cpp

```diff
@@ -88,37 +88,45 @@ void ShadowMapManager::render(FrameGraph& fg, FEngine& engine, FView& view,
     const TextureRequirements textureRequirements = mTextureRequirements;
     assert_invariant(textureRequirements.layers <= MAX_SHADOW_LAYERS);
 
-    using ShadowPass = std::pair<const ShadowMapEntry*, RenderPass>;
-    auto passes = utils::FixedCapacityVector<ShadowPass>::with_capacity(MAX_SHADOW_LAYERS);
+    struct ShadowPass {
+        ShadowPass(ShadowMapEntry const* shadowmap, RenderPass const& pass) noexcept
+            : shadowmap(shadowmap), pass(pass) {
+        }
+        ShadowMapEntry const* shadowmap;
+        RenderPass pass;
+    };
+
+    auto passList = utils::FixedCapacityVector<ShadowPass>::with_capacity(MAX_SHADOW_LAYERS);
 
     // These loops fill render passes with appropriate rendering commands for each shadow map.
     // The actual render pass execution is deferred to the frame graph.
     // Directional, cascaded shadowmaps
     for (const auto& map : mCascadeShadowMaps) {
-        if (!map.hasVisibleShadows()) {
+        auto& range = view.getVisibleDirectionalShadowCasters();
+        if (!map.hasVisibleShadows() || range.empty()) {
             continue;
         }
-        map.getShadowMap().render(driver, view.getVisibleDirectionalShadowCasters(), pass, view);
-        passes.emplace_back(&map, pass);
+        auto& entry = passList.emplace_back(&map, pass);
+        map.getShadowMap().render(driver, range, &entry.pass, view);
     }
 
     // Spotlight shadowmaps
     for (size_t i = 0, c = mSpotShadowMaps.size(); i < c; i++) {
         const auto& map = mSpotShadowMaps[i];
-        if (!map.hasVisibleShadows()) {
+        auto& range = view.getVisibleSpotShadowCasters();
+        if (!map.hasVisibleShadows() || range.empty()) {
             continue;
         }
-        pass.setVisibilityMask(VISIBLE_SPOT_SHADOW_RENDERABLE_N(i));
-        map.getShadowMap().render(driver, view.getVisibleSpotShadowCasters(), pass, view);
-        pass.clearVisibilityMask();
-        passes.emplace_back(&map, pass);
+        auto& entry = passList.emplace_back(&map, pass);
+        entry.pass.setVisibilityMask(VISIBLE_SPOT_SHADOW_RENDERABLE_N(i));
+        map.getShadowMap().render(driver, range, &entry.pass, view);
```

Comment:
Should we still reset the visibility mask after rendering the shadowmap?

Reply:
No because `emplace_back` creates a new `RenderPass` each time.